### PR TITLE
Add aya_vision

### PR DESCRIPTION
**Pull request by Minestar6**
# Add aya_vision
添加了aya_vision模型的PaddlePaddle 实现，包含模型代码、配置文件、单元测试。
# 对齐验证
## 精度对齐
前 20 个 token 的 logits 对齐结果如下：
* mean diff:  1.119111e-05
* max diff:  1.258850e-04

### 训练数据
文本数据
* GSM8K SFT 数据
* 训练步数：300 steps

多模态数据
### 训练结果
文本结果
2_ayavision
2_ayavision_diff

多模态结果
3_ayavision
3_ayavision_diff

### CODE REVIEW

File: paddleformers/transformers/aya_vision/modeling.py [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+            input_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            attention_mask=attention_mask,
+            logits_to_keep=logits_to_keep,
+            is_first_iteration=is_first_iteration,
+            **kwargs,
+        )
+        if is_first_iteration or not kwargs.get("use_cache", True) or past_key_values is None:
+            model_inputs["pixel_values"] = pixel_values
```
Comment (risemeup1111) [inline paddleformers/transformers/aya_vision/modeling.py:564]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这里在首个 generation step 会继续传入 `pixel_values`，但当前类没有覆盖 PaddleFormers 实际调用的 `expand_inputs_for_generation`。默认实现只会展开 `input_ids/attention_mask/position_ids` 等文本 batch 张量，不会展开 `pixel_values`；当 `num_beams>1` 或 `num_return_sequences>1` 时，`input_ids` 中的图像 token 数会被放大，而视觉特征仍是原 batch 数量，随后 `_merge_image_features()` 会因为 `n_image_tokens != flat_image_features.shape[0]` 抛出不匹配错误。这个 PR 新增的 beam 生成测试路径也会触发该问题。

请为 AyaVision 增加与当前生成入口匹配的视觉输入展开逻辑；如果继续支持 `crop_to_patches=True`，需要按样本的 tile 数切分后重复，而不是直接把整批 `pixel_values` 当作文本 batch gather：

```python
def expand_inputs_for_generation(self, input_ids, expand_size, attention_mask=None, **model_kwargs):
    source_input_ids = input_ids
    input_ids, model_kwargs = super().expand_inputs_for_generation(
        input_ids, expand_size=expand_size, attention_mask=attention_mask, **model_kwargs
    )
    pixel_values = model_kwargs.get("pixel_values")
    if expand_size > 1 and pixel_values is not None:
        # 根据 source_input_ids 中每个样本的图像 token 数和每个 tile 的 token 数计算切片长度，
        # 然后把每个样本对应的 pixel_values 切片 repeat expand_size 次。
        model_kwargs["pixel_values"] = repeat_visual_samples(pixel_values, lengths, expand_size)
    return input_ids, model_kwargs
```

File: paddleformers/transformers/aya_vision/modeling.py [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+        self.model = AyaVisionModel(config)
+        for attr in (
+            "sequence_parallel",
+            "tensor_parallel_output",
+            "use_filtered_label_loss",
+        ):
+            if hasattr(config, attr):
+                setattr(config.text_config, attr, getattr(config, attr))
+        self.lm_head = GeneralLMHead(config.text_config)
+        self.criterion = AyaVisionCriterionLayer(config)
```
Comment (risemeup1111) [inline paddleformers/transformers/aya_vision/modeling.py:449]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

`AyaVisionConfig`/`Cohere2Config` 默认 `tie_word_embeddings=True`，并且本类声明了 `_tied_weights_keys`，但普通构造路径只创建 `lm_head` 后没有调用 `tie_weights()`；目前只有 `from_pretrained()` 后置调用会绑定。这样 `AyaVisionForConditionalGeneration(config)` 用于从配置训练、单测或保存前推理时会保持未绑定输出头，和配置/检查点语义不一致。建议在初始化完 `lm_head/criterion` 后按配置绑定：

```suggestion
        self.criterion = AyaVisionCriterionLayer(config)
        if config.tie_word_embeddings:
            self.tie_weights()
```

File: paddleformers/transformers/aya_vision/processor.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        self.start_of_img_token = start_of_img_token
+        self.end_of_img_token = end_of_img_token
+        self.img_patch_token = img_patch_token
+        self.img_line_break_token = img_line_break_token
+        self.tile_token = tile_token
+        self.tile_global_token = tile_global_token
+        self.image_token_id = tokenizer.convert_tokens_to_ids(self.img_patch_token)
+        self.image_ids = tokenizer.convert_tokens_to_ids(
+            [img_patch_token, tile_token, tile_global_token, start_of_img_token, end_of_img_token]
+        )
```
Comment (risemeup1111) [inline paddleformers/transformers/aya_vision/processor.py]: P2 **优先级：P2**
**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

这里把完整的图像特殊 token 列表保存到了 `image_ids`，但仓库内没有读取这个属性；下面 `return_mm_token_type_ids=True` 调用的 `create_mm_token_type_ids()` 依赖的是 `image_token_ids` 属性。结果是 `TILE_GLOBAL`、起止图像 token 等不会按这里的列表参与 `mm_token_type_ids` 标记，开启该选项时多模态 token 类型会漏标。请改成 helper 实际使用的属性名，并补一个覆盖 `return_mm_token_type_ids=True` 的 processor 测试。

```suggestion
        self.image_token_ids = tokenizer.convert_tokens_to_ids(
            [img_patch_token, tile_token, tile_global_token, start_of_img_token, end_of_img_token]
        )
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查当前提交，CI 通过，但仍有会影响 AyaVision 生成路径和权重绑定语义的阻塞问题；另外有一个 processor 的低优先级问题需要作者回复。具体细节已放在行级评论中，请先按评论处理后再继续推进。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/aya_vision/processor.py (hunk 3 above)
Comment (risemeup1111) [reply paddleformers/transformers/aya_vision/processor.py]: 这次提交已经把 `image_token_ids` 和 `return_mm_token_type_ids=True` 的测试补上，代码方向我看是合理的。按这条 P2 的处理要求，还请作者在该线程回复 `Done`（或说明不同意见），收到回复后这项就可以关闭。

**Review by risemeup1111 — COMMENTED** [review]
已复查新提交：上轮两个 P1 问题已通过代码修复，相关补充测试也已加入。剩余的是上轮 P2 线程需要作者按评论要求回复 `Done` 或说明理由；我已在原线程补充说明。当前还有多项 CI 正在运行，建议等检查完成后再继续审批。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/aya_vision/processor.py (hunk 3 above)
Comment (Minestar6) [reply paddleformers/transformers/aya_vision/processor.py]: Done

Comment (risemeup1111) [reply paddleformers/transformers/aya_vision/processor.py]: 收到，已确认这条 P2 的代码修改和补充测试，回复满足处理要求。

**Review by risemeup1111 — COMMENTED** [review]
已确认作者在原 P2 线程回复 `Done`，上轮 P1/P2 评审项均已处理，当前没有新增代码行级问题。

P1 **优先级：P1**
非行级：CI 检查失败，无法挂载到具体 diff 行。当前 `Integration test (H20, single card)` 已失败，失败步骤为 `Qwen3-vl-8k-single-card`，同时还有若干检查仍在运行；失败日志暂因 workflow 未完全结束而不可用。请先确认并处理/重跑该 CI，等检查通过后再继续审批。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/aya_vision/configuration.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        super().__init__(**kwargs)
+        if isinstance(vision_config, dict):
+            vision_config["model_type"] = vision_config.get("model_type", "siglip_vision_model")
+            vision_config = CONFIG_MAPPING[vision_config["model_type"]](**vision_config)
+        elif vision_config is None:
+            vision_config = CONFIG_MAPPING["siglip_vision_model"](
+                hidden_size=1152,
+                intermediate_size=4304,
+                patch_size=14,
+                image_size=384,
```
Comment (risemeup1111) [inline paddleformers/transformers/aya_vision/configuration.py]: P1 `AyaVisionConfig()` 默认把 SigLIP `image_size` 设成 384，但 processor/mm_plugin 默认按 364 生成和预处理图像；默认 364 输入会产生 26x26 patch embeddings，却去加 27x27 position ids，视觉 forward 会直接形状不匹配。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查当前 head：上轮行级问题已处理；新发现默认视觉尺寸配置会阻断默认 AyaVision 视觉输入路径，已在行级评论标出。修复后也需要等当前仍在运行的 CI 完成。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by risemeup1111** [comment]
### Nyanpasu Review Board


| 序号 | 位置 | 优先级 | 状态 |
| - | - | - | - |
| 1 | [`paddleformers/transformers/aya_vision/modeling.py`](https://github.com/PaddlePaddle/PaddleFormers/pull/4720#discussion_r3610282874) | P1 | ✅ |
| 2 | [`paddleformers/transformers/aya_vision/modeling.py`](https://github.com/PaddlePaddle/PaddleFormers/pull/4720#discussion_r3610283378) | P1 | ✅ |
| 3 | [`paddleformers/transformers/aya_vision/processor.py`](https://github.com/PaddlePaddle/PaddleFormers/pull/4720#discussion_r3610283910) | P2 | ✅ |
| 4 | [非行级：CI 检查](https://github.com/PaddlePaddle/PaddleFormers/pull/4720#pullrequestreview-4744995324) | P1 | ✅ |
| 5 | [`paddleformers/transformers/aya_vision/configuration.py`](https://github.com/PaddlePaddle/PaddleFormers/pull/4720#discussion_r3675131979) | P1 | 🚧 |

<details>
<summary>点击展开 Review 规则</summary>
PR 评审规则：

- P0、P1 级别的评审必须提交新的 commit 进行修改；
- P2、P3 级别的评审可以通过评论进行修改。

状态说明：

| 已解决 | 无需修复 | 待修复 |
| - | - | - |
| ✅ | 🟡 | 🚧 |
</details>

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/cohere2/modeling.py [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+        self.is_causal = True
+        self.q_proj = nn.Linear(self.hidden_size, self.num_heads * self.head_dim, bias_attr=config.attention_bias)
+        self.k_proj = nn.Linear(
+            self.hidden_size, self.num_key_value_heads * self.head_dim, bias_attr=config.attention_bias
+        )
+        self.v_proj = nn.Linear(
+            self.hidden_size, self.num_key_value_heads * self.head_dim, bias_attr=config.attention_bias
+        )
+        self.o_proj = nn.Linear(self.num_heads * self.head_dim, self.hidden_size, bias_attr=config.attention_bias)
+        if config.use_qk_norm:
```
Comment (WYB27) [inline paddleformers/transformers/cohere2/modeling.py:96]: HF的实现没有qk_norm，官方权重也没有包含qk_norm的权重，是否可以考虑删掉相关逻辑

File: paddleformers/transformers/cohere2/modeling.py [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+            return_dict=return_dict,
+        )
+        hidden_states = outputs[0]
+        slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
+        hidden_states = hidden_states[:, slice_indices, :]
+        logits = self.lm_head(hidden_states)
+        logits = logits * self.logit_scale
+        loss = None
+        if labels is not None:
+            shift_logits = logits[:, :-1, :]
```
Comment (WYB27) [inline paddleformers/transformers/cohere2/modeling.py:521]: label移位已经在数据侧完成了吧，这里不用再移位了。另外下面的loss不走统一的CriterionLayer，而是手写实现的原因是？
